### ISSUE TITLE
Custom Environment Variables and JSON values

### ISSUE THREAD

**Issue by leachiM2k** [lorenwest/node-config#211]
Hey Loren,
our NODE_CONFIG environment variable has grown and is too big for our PaaS provider.
There is a great solution in node-config to specify custom environment variables (coming from #129).
But this solution has a drawback, because it does not support JSON in custom variables like it does in NODE_CONFIG.
This does not allow us to save space in NODE_CONFIG environment variable.
I would like to propose to add an JSON parsing to custom environment variables. This should be optional to keep backward compatibility.

If you like this idea, I would create a PR.

**Comment by lorenwest** [comment]
This sounds reasonable.  How would you indicate an environment variable needed JSON parsing?

**Comment by leachiM2k** [comment]
I thought of two possible solutions:

- As of today it is only allowed to specify strings in an environment variable. Therefore we could simply try to parse the variable's value with JSON.parse and take the original value if it fails.
- Indicate json content for example by appending "_JSON" to the variable's name. In my eyes it's a little bit ugly, but it keeps the backward compatibility without any possible side effects.

**Comment by lorenwest** [comment]
It seems like the `environment-variables.json` file contains instructions for environment variables, and that's the place we can put instructions for parsing JSON.  Possibly a new section?  What do you think?

**Comment by leachiM2k** [comment]
The file `custom-environment-variables.json` contains only the mapping for enviroment variables, actually it is a string at some point of the configuration tree. To enhance this information, it would be necessary to break the string into an object. It is possible, but I like the actual simple structure of this file.

Following the whole simplicity of node-config and the convention-over-configuration concept in node-config I would like to see my environment variables just to be parsed if they are valid JSON. Without any special format configuration.

What solution do you prefer?

**Comment by lorenwest** [comment]
I like convention over configuration, but I like backwards compatibility even more.  The problem with adding convention is it breaks installs for those un-knowingly using that convention.  You can make the convention esoteric and reduce the chances that someone will run up against it, but you can't eliminate the chances, and esoteric conventions aren't as valuable as obvious conventions.

With a zero tolerance for backward incompatibility, we'll either have to wait until a major version upgrade, or favor configuration over convention.  Which would you choose?

**Comment by leachiM2k** [comment]
Waiting for a major upgrade is of course no option for me.

I see actually three possible solutions.
* Changing `custom-environment-variables.json`from
```
{
  "Customer": {
    "dbConfig": {
      "host": "PROD_SERVER"
    }
  }
}
```
to
```
{
  "Customer": {
    "dbConfig": {
      "host": {"environmentVariable": "PROD_SERVER", "format": "json"}
    }
  }
}
```
**Code must support both to avoid backward incompatibility, probably not easy to read**

* Changing the name of the environment variable from SOMETHING to SOMETHING_JSON to get it parsed
**Near zero backward incompability, but a bit esoteric and still possible to have existing environment name collisions**

* Just try to parse the environment variable, if it fails take the unparsed value for compatibility reasons. `SOMETHING='http://example.com'` and `SOMETHING='{"port":3306,"server":"example.com"}'` are valid. First one is deprecated.
**Near zero backward incompability and feels organic, but still possible to hit an explicitly configured unparsed JSON string**

I like solution 3 very much, but in favor of full backward compatibility I would take solution 2.

Do you see a fourth solution?

**Comment by lorenwest** [comment]
I believe solution #1 is the only one that can guarantee backwards compatibility.  As it is now, they can only add strings as the value of the `host` variable (for example).  We can check to see if it's a string or an object, and if it's an object then it contains `environmentVariable` and `format` elements, where `format` can be one of our supported formats (json, js, yaml, etc.).

That delivers this functionality without breaking our commitment to backwards compatibility.

**Comment by markstos** [comment]
@leachiM2k regarding a major version upgrade, you always make an agreed-on change in a forward-compatible fork which you could start using immediately, so you would get the benefits of a major-upgrade change immediately, with waiting. 

Personally, I don't think backwards compatibility should be a priority. I've seen good software turn into bloated pile of hacked add-ons in the name of backcompat. What happens next is predictable: Someone else learns from the mistakes and starts over with a new project without all the bloat, which is unsurprisingly preferred by developers.

npm and semvar are great about not automatically upgrading people to incompatible versions.   

The Node.js and npm userbase is growing rapidly, so it's easy to believe that the userbase of this module will more than double in the future. That is, if the goal is satisfy the majority of our users, focusing on future users rather than past users should be the priority.

While the changes being discussed are breaking, they would require relatively minor changes for users wishing to upgrade to the major version bump.

**Comment by leachiM2k** [comment]
Hey back and thank you for this discussion so far.
I did two commits on our fork. Please let's have a look at these.

* The first commit https://github.com/lorenwest/node-config/commit/163d70074ef20690c4602fdf48f1eb586350812d introduces my preferred but at a point slightly backwards incompatible solution that just tries to parse the environment variable.

* The second commit (together with the first) https://github.com/lorenwest/node-config/compare/master...fid-dev:master is a follow-up and introduces @lorenwest's preferred solution that introduces two configuration parameters in `custom-environment-variables.EXT`, namely `__name` and `__format`.
The underscores are used to show the special meaning of the key and to avoid name collision in existing configs.

Please let me know what you think about it.

**Comment by dhm116** [comment]
:+1: 

Please tag + release this!

**Comment by jplock** [comment]
+1 please!

**Comment by mcholste** [comment]
+1!

**Comment by russellballestrini** [comment]
+1, I desire to pass in a single JSON document which holds details of the database server as a single ENV var, instead of chopping the JSON up into separate ENV vars and passing them separately.

**Comment by markstos** [comment]
Thanks for the feedback. It was curious to get four +1s within 10 minutes. Are you all using it for the same project?

I see the appeal of storing the various parts of a DB connection in a single environment variable.

**Comment by lorenwest** [comment]
Thanks for the nudge.  I'll get this released today.

-Loren

**Comment by lorenwest** [comment]
Published in config@1.14.0
